src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -57,7 +57,27 @@ private void ReturnReadBufferIfEmpty()
 
         ~SslStreamInternal()
         {
-            if (_internalBuffer != null)
+            Dispose(disposing: false);
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+
+            if (_internalBuffer == null)
+            {
+                // Suppress finalizer if the read buffer was returned.
+                GC.SuppressFinalize(this);
+            }
+        }
+
+        private void Dispose(bool disposing)
+        {
+            if (_internalBuffer != null &&
+                // When disposing, ensure a Read operation is not in progress,
+                // block potential reads since Sslstream is disposing.
+                ((disposing && Interlocked.Exchange(ref _nestedRead, 1) == 0) ||
+                !disposing))
```

Comment:
Looks like this could be simplified to !disposing || Interlocked.Exchange(...)...